### ISSUE TITLE
subgraph generates different results from cluster loader

### ISSUE THREAD

**Issue by llan-ml** [pyg-team/pytorch_geometric#7153]
### 🐛 Describe the bug

```ipython
In [1]: import torch

In [2]: from torch_geometric.datasets import Reddit

In [3]: from torch_geometric.loader import ClusterData, ClusterLoader

In [4]: K, N = 15, 100

In [5]: dataset = Reddit("./data/Reddit")

In [6]: data = dataset[0]

In [7]: cluster_data = ClusterData(data, num_parts=N, recursive=False, save_dir=dataset.processed_dir)

In [8]: batch = torch.tensor(range(K))

In [9]: start = cluster_data.partptr[batch].tolist()

In [10]: end = cluster_data.partptr[batch + 1].tolist()

In [11]: node_idx = torch.cat([torch.arange(s, e) for s, e in zip(start, end)])

In [12]: d = data.subgraph(node_idx)

In [13]: train_loader = ClusterLoader(cluster_data, batch_size=K, shuffle=False, num_workers=0)

In [14]: dd = train_loader._collate(batch)

In [15]: d
Out[15]: Data(x=[35019, 602], edge_index=[2, 2825258], y=[35019], train_mask=[35019], val_mask=[35019], test_mask=[35019])

In [16]: dd
Out[16]: Data(x=[35019, 602], y=[35019], train_mask=[35019], val_mask=[35019], test_mask=[35019], edge_index=[2, 14892264])
```

The number of edges of `d` is different from that of `dd`.

### Environment

* PyG version: 2.3.0
* PyTorch version: 1.13.1+cpu
* OS: Ubuntu 18.04
* Python version: 3.9.12
* CUDA/cuDNN version: N/A
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* Any other relevant information (*e.g.*, version of `torch-scatter`): torch-scatter==2.1.1+pt113cpu, torch-sparse==0.6.17+pt113cpu

**Comment by rusty1s** [comment]
That's because `ClusterData` reshuffles the data. As such, `partptr[i]:partptr[i+1]` does not refer to the original ordering. Therefore, using these indices in `subgraph` will reference different nodes.

I added a test in https://github.com/pyg-team/pytorch_geometric/pull/7159 to confirm that it indeed works as expected. PTAL.

**Comment by llan-ml** [comment]
Hi @rusty1s , thanks for your reply, and using the shuffled `cluster_data.data` to extract subgraphs yields the expected results.

In addition, I found that the speed of `subgraph` is several times faster than the current implementation of `ClusterLoader`. I tested on CPU and is not sure if this happens on GPU as well. 

Maybe we can use the function `subgraph` in the `_collate` of `ClusterLoader` like the [implementation in dgl](https://docs.dgl.ai/en/0.8.x/_modules/dgl/dataloading/cluster_gcn.html#ClusterGCNSampler).

**Comment by rusty1s** [comment]
Thanks for the hint. I assume the speedup is only noticeable on rather small graphs? Will look into.
